developer/src/keyman-vscode-plugin/src/kpjBuild.mts

```diff
@@ -0,0 +1,248 @@
+            ...options, readerOptions: {
+                importsPath: fileURLToPath(new URL(...LDMLKeyboardXMLSourceFileReader.defaultImportsURL))
+            }
+        };
+        const compiler = new kmcLdml.LdmlKeyboardCompiler();
+        if (!await compiler.init(callbacks, ldmlCompilerOptions)) {
+            msg(`Compiler failed init\r\n`);
+            continue;
+        }
+
+        const outfile = project.resolveOutputFilePath(path, KeymanFileTypes.Source.LdmlKeyboard, KeymanFileTypes.Binary.Keyboard);
+        msg(`.. outfile is ${outfile}\r\n`);
+        await mkParentDirs(outfile);
+        const result = await compiler.run(filePath, outfile);
+        if (!result) {
+            msg(`Compiler failed to run\r\n`);
+            continue;
+        }
+        msg(`.. compiled\r\n`);
+        // if(!this.createOutputFolder(outfile ?? infile, callbacks)) {
+        //             return false;
+        // }
+
+        if (!await compiler.write(result.artifacts)) {
+            msg(`Error writing ${outfile}\r\n`);
+            throw Error(`Error writing ${outfile}`);
+        }
+
+        msg(`.. wrote\r\n`);
+
+        msg(`\r\n\r\n`);
+        didCompileSrc = true; // we allow more than one xmk in each package
+    }
+
+    // now, compile any .kmn
+    for (const path of project.files.filter(({ filePath }) => extname(filePath) === KeymanFileTypes.Source.KeymanKeyboard)) {
+        const { filePath } = path;
+        msg(`Compiling KMN: ${filePath}\r\n`);
+
+        const compiler = new kmcKmn.KmnCompiler();
+        if (!await compiler.init(callbacks, coptions)) {
+            msg(`Compiler failed init\r\n`);
+            continue;
+        }
+
+        const outfile = project.resolveOutputFilePath(path, KeymanFileTypes.Source.KeymanKeyboard, KeymanFileTypes.Binary.Keyboard);
+        msg(`.. outfile is ${outfile}\r\n`);
+        const infilePosix = getPosixAbsolutePath(filePath);
+        const outfilePosix = getPosixAbsolutePath(outfile);
+        await mkParentDirs(outfilePosix);
+        msg(`${infilePosix} => ${outfilePosix}\r\n`);
+        const result = await compiler.run(infilePosix, outfilePosix);
+        if (!result) {
+            msg(`Compiler failed to run\r\n`);
+            continue;
+        }
+        msg(`.. compiled\r\n`);
+        // if(!this.createOutputFolder(outfile ?? infile, callbacks)) {
+        //             return false;
+        // }
+
+        if (!await compiler.write(result.artifacts)) {
+            msg(`Error writing ${outfile}\r\n`);
+            throw Error(`Error writing ${outfile}`);
+        }
+
+        msg(`.. wrote\r\n`);
+
+        msg(`\r\n\r\n`);
+        didCompileSrc = true; // we allow more than one xmk in each package
+    }
+
+    // check errs and get out
+
+    if(callbacks.hasFailureMessage(false)) {
+        throw Error(`Error building ${kpjPath}`);
+    }
+
+    if (!didCompileSrc) {
+        throw Error(`Error: no source files were compiled.`);
+    }
```

Comment:
It is legitimate to have a project with only a .kps and no .kmn or .xml

Reply:
- [ ] remove error